### DISCUSSION TITLE
Multiple (not repetitive) dosing

### DISCUSSION THREAD

**Discussion by PascaleRietveld** [humanpred/pknca#242] [Q&A] [answered]
---

Hi Bill,

Thank you for your useful tutorial! I am trying to calculate the AUC of my population but because of multiple dosing (not repetitive) I cannot make it seem to work.
I have a dataset where each patient receives per cycle (2 cycles per ID) an IV bolus and then a continuous infusion of 22 hours or 44 hours. Then some patients receive a second bolus and continuous infusion after about 22 hours. I use time after first dose (TAFD) because of multiple dosing.
My dataset kindoff looks like this:

 

 
```r

Data <- data.frame(
  cycle = c(1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 1, 1),
  ID = c(1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2),
  Conc = c(0, 0, 2.8, 2.5, 0, 0, 0, 0, 3, 3.2, 0, 0, 5),
  TAFD = c(0, 0.4, 5, 22, 0, 0.5, 0, 0.5, 10, 22, 23, 23.5, 45),
  duration = c(0.25, 22, NA, NA, 0.25, 22, 0.25, 22, NA, NA, 0.25, 22, NA),
  Dose = c(700, 1000, NA, NA, 500, 1100, 700, 1300, NA, NA, 700, 1100, NA))
 
data_conc <- Data %>% filter(is.na(Dose)) %>% select(-Dose,-duration)
 
knitr::kable(Data)
 
 
 
my.conc <- PKNCAconc(data_conc, Conc~TAFD|ID)
 
Dosdata <- Data %>% select(ID, Dose, TAFD, duration) %>% filter(!is.na(Dose))
my.dose <- PKNCAdose(Dosdata, Dose~TAFD|ID)
```
I am stuck because of this error: Rows that are not unique per group and time (column names: TAFD, ID) found within dosing data.

How can I calculate the PK parameters for this multiple dosing dataset?
Thanks a lot in advance!

Best,
Pascale

**Comment by billdenney** [comment, marked as answer]
@PascaleRietveld, Thanks for using PKNCA.  I've moved this to a discussion since it's a question about use rather than a problem with PKNCA.

Here is your original data/code as a reprex showing the specific error and with some minor modifications to highlight the issue.  From your code, I added the `cycle` column to `Dosdata`, and I added a line to show the `Dosdata` at the bottom:

``` r
library(tidyverse)
#> Warning: package 'ggplot2' was built under R version 4.3.1
#> Warning: package 'purrr' was built under R version 4.3.1
library(PKNCA)
#> 
#> Attaching package: 'PKNCA'
#> The following object is masked from 'package:stats':
#> 
#>     filter

Data <- data.frame(
  cycle = c(1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 1, 1),
  ID = c(1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2),
  Conc = c(0, 0, 2.8, 2.5, 0, 0, 0, 0, 3, 3.2, 0, 0, 5),
  TAFD = c(0, 0.4, 5, 22, 0, 0.5, 0, 0.5, 10, 22, 23, 23.5, 45),
  duration = c(0.25, 22, NA, NA, 0.25, 22, 0.25, 22, NA, NA, 0.25, 22, NA),
  Dose = c(700, 1000, NA, NA, 500, 1100, 700, 1300, NA, NA, 700, 1100, NA))

data_conc <- Data %>% filter(is.na(Dose)) %>% select(-Dose,-duration)

my.conc <- PKNCAconc(data_conc, Conc~TAFD|ID)

Dosdata <- Data %>% select(ID, Dose, TAFD, duration, cycle) %>% filter(!is.na(Dose))
my.dose <- PKNCAdose(Dosdata, Dose~TAFD|ID)
#> Error in PKNCAdose.data.frame(Dosdata, Dose ~ TAFD | ID): Rows that are not unique per group and time (column names: TAFD, ID) found within dosing data.  Row numbers: 3

Dosdata
#>   ID Dose TAFD duration cycle
#> 1  1  700  0.0     0.25     1
#> 2  1 1000  0.4    22.00     1
#> 3  1  500  0.0     0.25     2
#> 4  1 1100  0.5    22.00     2
#> 5  2  700  0.0     0.25     1
#> 6  2 1300  0.5    22.00     1
#> 7  2  700 23.0     0.25     1
#> 8  2 1100 23.5    22.00     1
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

The error is saying that row 3 of your data is duplicated for the dosing data.  The reason it's duplicated is because `TAFD` and `ID` are the same on that row compared to another (in this case, the duplicated row is row 1).

The way to clarify the difference between rows 1 and 3 in your data is to include `cycle` in the grouping (see `cycle+ID` for both `PKNCAconc()` and `PKNCAdose()` below):

``` r
library(tidyverse)
#> Warning: package 'ggplot2' was built under R version 4.3.1
#> Warning: package 'purrr' was built under R version 4.3.1
library(PKNCA)
#> 
#> Attaching package: 'PKNCA'
#> The following object is masked from 'package:stats':
#> 
#>     filter

Data <- data.frame(
  cycle = c(1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 1, 1),
  ID = c(1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2),
  Conc = c(0, 0, 2.8, 2.5, 0, 0, 0, 0, 3, 3.2, 0, 0, 5),
  TAFD = c(0, 0.4, 5, 22, 0, 0.5, 0, 0.5, 10, 22, 23, 23.5, 45),
  duration = c(0.25, 22, NA, NA, 0.25, 22, 0.25, 22, NA, NA, 0.25, 22, NA),
  Dose = c(700, 1000, NA, NA, 500, 1100, 700, 1300, NA, NA, 700, 1100, NA))

data_conc <- Data %>% filter(is.na(Dose)) %>% select(-Dose,-duration)

my.conc <- PKNCAconc(data_conc, Conc~TAFD|cycle+ID)

Dosdata <- Data %>% select(ID, Dose, TAFD, duration, cycle) %>% filter(!is.na(Dose))
my.dose <- PKNCAdose(Dosdata, Dose~TAFD|cycle+ID)
#> Found column named duration, using it for the attribute of the same name.
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Another option would be to make the `TAFD` column represent the time since the first dose in the first cycle; then there would be no duplication.

**Reply by PascaleRietveld** [reply]
Thank you Bill! I am clearly new to this:)
Later down the code I get the message to report a bug.
The PKNCAdata function also gives me trouble that I don't understand.
What I see is this:

>use the default settings for AUC ranges, whether to get cmax, tmax, kel.pred, etc.)
my.data.auto <- PKNCAdata(my.conc,my.dose)
Error in vectbl_assign(x[[j]], i, recycled_value[[j]]) : 
  DLL requires the use of native symbols

PKNCA.options("single.dose.aucs")
my.data.auto$intervals


manually set what items to calculate and what intervals
my.intervals <- data.frame(start = 0,
                           end=50,
                           cmax=T,
                           tmax=T,
                           aucinf=F,
                           auclast=T,
                           kel.obs=T,
                           cl.obs=T)

my.data.manual <- PKNCAdata(my.conc,my.dose, intervals=my.intervals)
my.data.manual$intervals



=============== Results ======================
auto
>my.results.auto <- pk.nca(my.data.auto)
Error in `purrr::pmap()`:
ℹ In index: 1.
Caused by error in `vectbl_assign()`:
! Please report a bug.
: Error with interval start=0.416666666666667, end=Inf: DLL requires the use of native symbols

Should I change column names or types to use native symbols?

Best,
Pascale

**Reply by billdenney** [reply]
The fact that the error is with `vectbl_assign` suggests that the underlying issue is with the `vctrs` library (see tidyverse/dplyr#6793).

Can you please provide the output of `sessionInfo()`?

**Reply by PascaleRietveld** [reply]
Hi Bill,

Somehow the error went away (I think because I work on a different computer today). 
Everything worked well in the end, using the data imputation method start_conc0.
Thank you!

**Reply by billdenney** [reply]
Thanks for following up.  I think that the underlying issue you were seeing the second time was with the `vctrs` library and that's why a different computer (possibly with a different version of some other libraries) worked.
